utils/litemap/src/store/vec_impl.rs

```diff
@@ -104,38 +104,170 @@ impl<K, V> StoreMut<K, V> for Vec<(K, V)> {
+    // This operation is still `O(n)`.
+    //
+    // Example:
+    // Assume T is (char, u8) and the bucketing function is equality on the char.
+    //
+    // We start in this state, where `r` represents "next
+    // read" and `w` represents "next_write".
+    //
+    //              r
+    //     | a,0 | b,0 | b,1 | c,0 | d,0 | d,1 |
+    //             w
+    //
+    // Comparing self[r] against self[w-1], this is not a duplicate, so
+    // we swap self[r] and self[w] (no effect as r==w) and then increment both
+    // r and w, leaving us with:
+    //
+    //                    r
+    //     | a,0 | b,0 | b,1 | c,0 | d,0 | d,0 |
+    //                    w
+    //
+    // Comparing self[r] against self[w-1], this value is a duplicate,
+    // we swap self[r] and self[w-1] and then increment `r`:
+    //
+    //                          r
+    //     | a,0 | b,1 | b,0 | c,0 | d,0 | d,1 |
+    //                   w
+    //
+    // Comparing self[r] against self[w-1], this is not a duplicate,
+    // so swap self[r] and self[w] and advance r and w:
+    //
+    //                                r
+    //     | a,0 | b,1 | c,0 | b,0 | d,0 | d,1 |
+    //                          w
+    //
+    // Comparing self[r] against self[w-1], this is not a duplicate,
+    // so swap self[r] and self[w] and advance r and w:
+    //
+    //                                      r
+    //     | a,0 | b,1 | c,0 | d,0 | b,0 | d,1 |
+    //                                w
+    //
+    // Comparing self[r] against self[w-1], this value is a duplicate,
+    // we swap self[r] and self[w-1] and then increment `r`:
+    //                                             r
+    //     | a,0 | b,1 | c,0 | d,1 | b,0 | d,0 |
+    //                                w
+    //
+    // End of slice, as r > len. Split at w.
+
+    let len = v.len();
+    if len <= 1 {
+        return (v, &mut []);
+    }
+
+    let ptr = v.as_mut_ptr();
+    let mut next_read: usize = 1;
+    let mut next_write: usize = 1;
+
+    // SAFETY: the `while` condition guarantees `next_read` and `next_write`
+    // are less than `len`, thus are inside `self`. `prev_ptr_write` points to
+    // one element before `ptr_write`, but `next_write` starts at 1, so
+    // `prev_ptr_write` is never less than 0 and is inside the slice.
+    // This fulfils the requirements for dereferencing `ptr_read`, `prev_ptr_write`
+    // and `ptr_write`, and for using `ptr.add(next_read)`, `ptr.add(next_write - 1)`
+    // and `prev_ptr_write.offset(1)`.
+    //
+    // `next_write` is also incremented at most once per loop at most meaning
+    // no element is skipped when it may need to be swapped.
+    //
+    // `ptr_read` and `prev_ptr_write` never point to the same element. This
+    // is required for `&mut *ptr_read`, `&mut *prev_ptr_write` to be safe.
+    // The explanation is simply that `next_read >= next_write` is always true,
+    // thus `next_read > next_write - 1` is too.
+    unsafe {
```

Comment:
Discussion: I think I'd prefer to not use `unsafe` here, because

1. We don't currently have any `unsafe` code in the `litemap` library crate. (I see some in integration tests but not in `src`.)
2. Everything else you're doing is already a big speedup; we don't need to hyper-optimize every last operation.
3. Bounds checks for this pattern, the "tortoise and hare" problem, are something that LLVM should eventually optimize. https://users.rust-lang.org/t/hare-and-tortoise-indexes-analysis/63985